<a href="https://colab.research.google.com/github/ali-homayouni/X-HSD/blob/master/Copy_of_HW2_FMNLP_2022.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Homework 2, FMNLP Fall 2022
### You have ten days for this assignment, please submit as PDF(File>Print>Save as PDF). 100 points total.

---
If you have any further questions or concerns, contact the TA via email:
borhanifard.zeinab@gmail.com or borhanifardz@ut.ac.ir
#####Telegram: @borhanifard

***LINK: *paste your link here****

---

#### IMPORTANT: After copying this notebook to your Google Drive, please paste a link to it below. To get a publicly-accessible link, hit the *Share* button at the top right, then click "Get shareable link" and copy over the result. If you fail to do this, you will receive no credit for this homework!
***LINK:***

---
##### *How to submit this problem set:*
- Write all the answers in this Colab notebook. Once you are finished, generate a PDF via (File -> Print -> Save as PDF) and upload it to Elearn (elearn.ut.ac.ir).
  
- **Important:** check your PDF before you submit to Elearn to make sure it exported correctly. If Colab gets confused about your syntax, it will sometimes terminate the PDF creation routine early.

- **Important:** on Elearn, please make sure that you tag each page with the corresponding question(s). This makes it significantly easier for our graders to grade submissions, especially with the long outputs of many of these cells. We will take off points for submissions that are not tagged.

- When creating your final version of the PDF to hand in, please do a fresh restart and execute every cell in order. One handy way to do this is by clicking `Runtime -> Run All` in the notebook menu.

---

##### *Academic honesty*

- We will audit the Colab notebooks from a set number of students, chosen at random. The audits will check that the code you wrote actually generates the answers in your PDF. If you turn in correct answers on your PDF without code that actually generates those answers, we will consider this a serious case of cheating. See the course page for honesty policies.

- We will also run automatic checks of Colab notebooks for plagiarism. Copying code from others is also considered a serious case of cheating.

---

# Part 1: Annotation

In this homework, you will first collect a labeled dataset of **120** **Persian sentences**  for a text classification task of your choice. **In order to complete Part 2 of this homework and obtain more training data, every three students must collect data on the same topic. There should be at least 360 Persian sentences in Part 2.** This process will include:

1. *Data collection*: Collect **120 Persian sentences** from any source you find interesting (e.g., literature, Tweets, news articles, reviews, etc.)

2. *Task design*: Come up with a binary (i.e., only two labels) sentence-level classification task that you would like to perform on your sentences. Be creative, and make sure your task isn't too easy (e.g., perhaps the labels have some degree of subjectivity to them, or the task is otherwise complex for humans). Write up annotator guidelines/instructions on how you would like people to label your data.

3. On your dataset, collect annotations from **two** classmates for your task. Everyone in this class will need to both create their own dataset and also serve as an annotator for two other classmates. In order to get everything done on time, you need to complete the following steps:

> *   Find two classmates willing to label 120 sentences each.
*   Send them your annotation guidelines and a way that they can easily annotate the data (e.g., a spreadsheet or Google form)
*   Collect the labeled data from each of the two annotators.
*   Sanity check the data for basic cleanliness (are all examples annotated? are all labels allowable ones?)

4. Collect feedback from annotators about the task including annotation time and obstacles encountered (e.g., maybe your guidelines were confusing! or maybe some sentences were particularly hard to annotate!)

5. Calculate and report inter-annotator agreement.

6. Aggregate output from both annotators to create final dataset.

7. Perform NLP experiments on your new dataset!

## Question 1.1 (10 points):
Describe the source of your unlabeled data, why you chose it, and what kind of sentence selection process you used (if any) to choose 120 sentences for annotation. Also briefly describe the text classification task that you will be collecting labels for in the next section.

### The chosen text classification task is Hate Speech detection in the Persian language. This task consists of two labels, hate, and no. In the definition of these labels, offensive speech is a converse that makes bad feelings for the reader or audience. Hate speech is offensive speech with a particular target community. 
### According to the context of my selected classification topic, we need to choose a data source consisting of offensive and hate speech. Also, the data source should be easily accessible without any content filter. So we decided to crawl data from Twitter. 
### The data selection process consists of two steps; first, fetching a bunch of tweets that contain at least one related keyword of the target group(These keywords were collected in a former project), then selecting the tweets that have at least one offensive word. Also selecting some data without any target or offensive keyword to consider as a non-hate class candidate.


## Question 1.2 (25 points):
Copy the annotation guidelines that you provided to your classmates below. We expect that these guidelines will be very detailed (and as such could be fairly long). You must include:

> *   The two categories for your binary classification problem, including the exact strings you want annotators to use while labeling. 
*   Descriptions of the categories and what they mean.
*   Representative examples of each category (i.e., sentences from outside your dataset that you have manually labeled to give annotators an idea of how to perform the task)
*   A discussion of of tricky corner cases, and criteria to help the annotator decide them. If you look at the data and think about how an annotator could do the task, you will likely find a bunch of these!


---


In general, offensive behavior and speech is a set of expressions and phrases that cause discomfort, resentment, anger or disgust in the audience or their readers. This type of behavior and speech is generally harsh, impolite, contains vulgar words and similar characteristics.

The definition of hate speech and behavior depends on the target community. However, many cases of public hatred are common in all societies. For example, racial hatred is a common hateful element
It is between all nations.
Here, we consider the set of hate speech and behavior specific to insulting the following groups. As we said before, these groups do not include all hate cases; Because our target society is Iranian society.
Hateful speech is any insulting speech that promotes and incites anger and hatred to people
groups with one of the following commonalities:

>*   Age
*   Guild
*   Disability
*   Race
*   Sex / Gender
*   Nationality
*   Immigration Status
*   Religion
*   Victims of a major violent event and their kin

According to the above definitions, we will have two general groups in labeling. The first group are expressions that are either not offensive or are only offensive and not hateful. The second group are expressions that are offensive.

### Labeling method:
In labeling, attention has been paid to two issues. First, some tweets are clearly offensive. Open insult 2 is an insult in which a person or persons are insulted by using a word or words. In contrast to this type of insult, we have implicit insult. Implicit insult is an insult in which a person insults a person or persons by using language expressions such as irony, metaphor and other language tools. Recognizing this type of insult and labeling it, unlike the first type of insult which is obvious, requires attention to the content of the text.

Some of the expressions used in themselves do not have an offensive meaning. These phrases appear in both explicit and implicit insults. In terminology, we call these expressions neutral expressions. The way these expressions are labeled is similar to implicit insult depending on the context.

Sometimes we come across insulting expressions that are widely used in society. For example, the use of the term "بیشعور" is abundantly seen in the speech of the elders of the society. In tagging, this frequency and how common the phrase is is not taken into account.

In the case of hatred towards religion, we consider insulting the text of a religion as only offensive content. However, contrary to the definition of hateful content, we consider insulting the character of a religion or its elders to be hateful. Because this type of insult hurts the feelings of the followers of that religion, which is considered a kind of insult and mockery of a group.

Examples: 

| Tweet                                                                                                                                                                                                       | Label | Description        |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------|--------------------|
| الان در حال حاضر منفورترین ها در جامعه فارسی دو بخش هستند. بخش اول مجاهدین خلق بخش دوم کاربر سعای حرومزاده ی هزار پدر                                                                                        | NO    | No Targer Group    |
| همینایی که تو افغانستان برای کشورشون نحنگیدن و از کشورشون فرار کردن، چند وقت دیگه که پناهندگی گرفتن یه روزی لب دریا کنار ساحل تو خیابون با دیدن خانومای بی حجاب یادشون میوفته که اسلام در خطره و باید اسلام | HATE  | Immigration Status |


## Question 1.3 (5 points):
Write down the names and emails of the two classmates who will be annotating your data below. Once they are finished annotating, create two .csv files (annotator1.csv and annotator2.csv) that contains each annotator's labels. The file should have two columns with headers **text** and **label**, respectively. You will include these files in an email to the instructors account when you're finished with this homework. 

*The tweets.csv file provided as an example in Part 2 below uses the same format.*

### *Hosein Seifi/hosein.seifi@ut.ac.ir* ###
### *Elham Khalaj/elham.khalaj@ut.ac.ir* ###

## Question 1.4 (10 points):
After both annotators have finished labeling the 120 sentences you gave them, ask them for feedback about your task and the provided annotation guidelines. If you were to collect more labeled data for this task in the future, what would you change from your current setup? Why? Please include a summary of annotator feedback (with specific examples that they found challenging to label) in your answer.

### *WRITE ANSWER HERE* ###

## Question 1.5 (10 points):
Now, compute the inter-annotator agreement between your two annotators. Upload both .csv files to your Colab session (click the folder icon in the sidebar to the left of the screen). In the code cell below, read the data from the two files and compute both the raw agreement (% of examples for which both annotators agreed on the label) and the [Cohen's Kappa](https://en.wikipedia.org/wiki/Cohen%27s_kappa). Feel free to use implementations in existing libraries (e.g., [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html)). After you're done, paste the numbers in the text cell that follows your code. 

*If you're curious, Cohen suggested the Kappa result be interpreted as follows: values ≤ 0 as indicating no agreement and 0.01–0.20 as none to slight, 0.21–0.40 as fair, 0.41– 0.60 as moderate, 0.61–0.80 as substantial, and 0.81–1.00 as almost perfect agreement.*

In [ ]:
import pandas as pd
from sklearn.metrics import cohen_kappa_score

d1 = pd.read_csv('./elham.csv')
d2 = pd.read_csv('./hosein.csv')

agreement = len(d1[d1['label'] == d2['label']]) / 120
cohen = cohen_kappa_score(d1['label'],d2['label'])

print('agreement:', agreement)
print('cohen:', cohen)

agreement: 0.9166666666666666
cohen: 0.8327759197324415


###*RAW AGREEMENT*: 0.9166666666666666
###*COHEN'S KAPPA*: 0.8327759197324415

## Question 1.6 (10 points):
To form your final dataset, you need to *aggregate* the annotations from both annotators (i.e., for cases where they disagree, you need to choose a single label). Use any method you like other than random label selection to perform this aggregation (e.g., have the two annotators discuss each disagreement and come to consensus, or choose the label you agree with the most). Upload your final dataset to the Colab session (in the same format as the other two files) as final_dataset.csv. Remember to include this file in your final email to us!

### *According to the annotation standards, in a disagreement situation, the best approach is to use majority labels. While we have two annotators and practically, it's not possible. So I annotated tweets, too, then used the majority to choose the final labels.* ###

# Part 2: Text classification

Now we'll move onto fine-tuning  pretrained language models specifically on your dataset. This part of the homework is meant to be an introduction to the HuggingFace library, and it contains code that will potentially be useful for your final projects. Since we're dealing with large models, the first step is to change to a GPU runtime.

## Adding a hardware accelerator

Please go to the menu and add a GPU as follows:

`Edit > Notebook Settings > Hardware accelerator > (GPU)`

Run the following cell to confirm that the GPU is detected.

In [ ]:
import torch

# Confirm that the GPU is detected

assert torch.cuda.is_available()

# Get the GPU device name.
device_name = torch.cuda.get_device_name()
n_gpu = torch.cuda.device_count()
print(f"Found device: {device_name}, n_gpu: {n_gpu}")
device = torch.device("cuda")

Found device: Tesla T4, n_gpu: 1


## Installing Hugging Face's Transformers library
We will use Hugging Face's Transformers (https://github.com/huggingface/transformers), an open-source library that provides general-purpose architectures for natural language understanding and generation with a collection of various pretrained models made by the NLP community. This library will allow us to easily use pretrained models like `BERT` and perform experiments on top of them. We can use these models to solve downstream target tasks, such as text classification, question answering, and sequence labeling.

Run the following cell to install Hugging Face's Transformers library and download a sample data file called tweets.csv that contains tweets about airlines along with a negative, neutral, or positive sentiment rating. Note that you will be asked to link with your Google Drive account to download some of these files. If you're concerned about security risks (there have not been any issues in previous semesters), feel free to make a new Google account and use it for this homework!

In [ ]:
!pip install transformers
!pip install -U -q PyDrive

from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
print('success!')

import os
import zipfile

# Download helper functions file
helper_file = drive.CreateFile({'id': '16HW-z9Y1tM3gZ_vFpJAuwUDohz91Aac-'})
helper_file.GetContentFile('helpers.py')
print('helper file downloaded! (helpers.py)')

# Download sample file of tweets
data_file = drive.CreateFile({'id': '1QcoAmjOYRtsMX7njjQTYooIbJHPc6Ese'})
data_file.GetContentFile('tweets.csv')
print('sample tweets downloaded! (tweets.csv)')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
success!
helper file downloaded! (helpers.py)
sample tweets downloaded! (tweets.csv)


The cell below imports some helper functions we wrote to demonstrate the task on the sample tweet dataset.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy

# Part 1: Data Prep and Model Specifications

Upload your data using the file explorer to the left. We have provided a function below to tokenize and format your data as BERT requires. Make sure that your csv file, titled final_data.csv, has one column "text" and another column "labels" containing integers.

If you run the cell below without modifications, it will run on the tweets.csv example data we have provided. It imports some helper functions we wrote to demonstrate the task on the sample tweet dataset. You should first run all of the following cells with tweets.csv just to see how everything works. Then, once you understand the whole preprocessing / fine-tuning process, change the csv in the below cell to your final_data.csv file, add any extra preprocessing code you wish, and then run the cells again on your own data.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy
import pandas as pd

df = pd.read_csv('final_data.csv')
# df = pd.read_csv('tweets.csv')

df = df.sample(frac=1).reset_index(drop=True)

texts = df.text.values
labels = df.label.values

### tokenize_and_format() is a helper function provided in helpers.py ###
input_ids, attention_masks = tokenize_and_format(texts)

# Convert the lists into tensors.
input_ids = torch.cat(input_ids, dim=0)
attention_masks = torch.cat(attention_masks, dim=0)
labels = torch.tensor(labels)

# Print sentence 0, now as a list of IDs.
print('Original: ', texts[0])
print('Token IDs:', input_ids[0])

Original:   bahar__irani ای خامنه ای بیناموس 👎👎👎👎👎😭😭😭😭😭
Token IDs: tensor([  101, 13253,  2099,  1035,  1035,  4238,  2072,  1270, 24830,  1277,
        25573, 22192, 15915, 14157,  1270, 24830,  1271, 24830, 15915, 25573,
        22192, 29836, 29824,   100,   102,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0])


## Create train/test/validation splits

Here we split your dataset into 3 parts: a training set, a validation set, and a testing set. Each item in your dataset will be a 3-tuple containing an input_id tensor, an attention_mask tensor, and a label tensor.



In [ ]:

total = len(df)

num_train = int(total * .8)
num_val = int(total * .1)
num_test = total - num_train - num_val

# make lists of 3-tuples (already shuffled the dataframe in cell above)

train_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train)]
val_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train, num_val+num_train)]
test_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_val + num_train, total)]

train_text = [texts[i] for i in range(num_train)]
val_text = [texts[i] for i in range(num_train, num_val+num_train)]
test_text = [texts[i] for i in range(num_val + num_train, total)]


Here we choose the model we want to finetune from https://huggingface.co/transformers/pretrained_models.html. Because the task requires us to label sentences, we wil be using BertForSequenceClassification below. You may see a warning that states that `some weights of the model checkpoint at [model name] were not used when initializing. . .` This warning is expected and means that you should fine-tune your pre-trained model before using it on your downstream task. See [here](https://github.com/huggingface/transformers/issues/5421#issuecomment-652582854) for more info.

In [ ]:
from transformers import BertForSequenceClassification, AdamW, BertConfig

model = BertForSequenceClassification.from_pretrained(
    f'arm-on/BERTweet-FA', # Use the 12-layer BERT model, with an uncased vocab.
    num_labels = 3, # The number of output labels.   
    output_attentions = False, # Whether the model returns attentions weights.
    output_hidden_states = False, # Whether the model returns all hidden-states.
)

# Tell pytorch to run this model on the GPU.
model.cuda()


# ACTION REQUIRED #

Define your fine-tuning hyperparameters in the cell below (we have randomly picked some values to start with). We want you to experiment with different configurations to find the one that works best (i.e., highest accuracy) on your validation set. Feel free to also change pretrained models to others available in the HuggingFace library (you'll have to modify the cell above to do this). You might find papers on BERT fine-tuning stability (e.g., [Mosbach et al., ICLR 2021](https://openreview.net/pdf?id=nzpLWnVAyah)) to be of interest.

In [ ]:
batch_size = 99
optimizer = AdamW(model.parameters(),
                  lr = 2.5e-5, # args.learning_rate - default is 5e-5
                  eps = 1e-15  # args.adam_epsilon  - default is 1e-8
                )
epochs = 15

/usr/local/lib/python3.7/dist-packages/transformers/optimization.py:310: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,


# Fine-tune your model
Here we provide code for fine-tuning your model, monitoring the loss, and checking your validation accuracy. Rerun both of the below cells when you change your hyperparameters above.

## collaps

In [ ]:
import numpy as np
# function to get validation accuracy
def get_validation_performance(val_set):
    # Put the model in evaluation mode
    model.eval()

    # Tracking variables 
    total_eval_accuracy = 0
    total_eval_loss = 0

    num_batches = int(len(val_set)/batch_size) + 1

    total_correct = 0

    for i in range(num_batches):

      end_index = min(batch_size * (i+1), len(val_set))

      batch = val_set[i*batch_size:end_index]
      
      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])
      
      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)
        
      # Tell pytorch not to bother with constructing the compute graph during
      # the forward pass, since this is only needed for backprop (training).
      with torch.no_grad():        

        # Forward pass, calculate logit predictions.
        outputs = model(b_input_ids, 
                                token_type_ids=None, 
                                attention_mask=b_input_mask,
                                labels=b_labels)
        loss = outputs.loss
        logits = outputs.logits
            
        # Accumulate the validation loss.
        total_eval_loss += loss.item()
        
        # Move logits and labels to CPU
        logits = logits.detach().cpu().numpy()
        label_ids = b_labels.to('cpu').numpy()

        # Calculate the number of correctly labeled examples in batch
        pred_flat = np.argmax(logits, axis=1).flatten()
        labels_flat = label_ids.flatten()
        num_correct = np.sum(pred_flat == labels_flat)
        total_correct += num_correct
        
    # Report the final accuracy for this validation run.
    avg_val_accuracy = total_correct / len(val_set)
    return avg_val_accuracy



## colapps

In [ ]:
import random

# training loop

# For each epoch...
for epoch_i in range(0, epochs):
    # Perform one full pass over the training set.

    print("")
    print('======== Epoch {:} / {:} ========'.format(epoch_i + 1, epochs))
    print('Training...')

    # Reset the total loss for this epoch.
    total_train_loss = 0

    # Put the model into training mode.
    model.train()

    # For each batch of training data...
    num_batches = int(len(train_set)/batch_size) + 1

    for i in range(num_batches):
      end_index = min(batch_size * (i+1), len(train_set))

      batch = train_set[i*batch_size:end_index]

      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])

      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)

      # Clear the previously calculated gradient
      model.zero_grad()        

      # Perform a forward pass (evaluate the model on this training batch).
      outputs = model(b_input_ids, 
                            token_type_ids=None, 
                            attention_mask=b_input_mask, 
                            labels=b_labels)
      loss = outputs.loss
      logits = outputs.logits

      total_train_loss += loss.item()

      # Perform a backward pass to calculate the gradients.
      loss.backward()

      # Update parameters and take a step using the computed gradient.
      optimizer.step()
        
    # ========================================
    #               Validation
    # ========================================
    # After the completion of each training epoch, measure our performance on
    # our validation set. Implement this function in the cell above.
    print(f"Total loss: {total_train_loss}")
    val_acc = get_validation_performance(val_set)
    print(f"Validation accuracy: {val_acc}")
    
print("")
print("Training complete!")



======== Epoch 1 / 15 ========
Training...


RuntimeError: ignored

# Evaluate your model on the test set
After you're satisfied with your hyperparameters (i.e., you're unable to achieve higher validation accuracy by modifying them further), it's time to evaluate your model on the test set! Run the below cell to compute test set accuracy.


In [ ]:
get_validation_performance(test_set)

RuntimeError: ignored

## Question 2.1 (10 points):
Congratulations! You've now gone through the entire fine-tuning process and created a model for your downstream task. Two more questions left :) First, describe your hyperparameter selection process in words. If you based your process on any research papers or websites, please reference them. Why do you think the hyperparameters you ended up choosing worked better than others? Also, is there a significant discrepancy between your test and validation accuracy? Why do you think this is the case?

### *WRITE YOUR ANSWER HERE*

## Question 2.2 (20 points):
Finally, perform an *error analysis* on your model. This is good practice for your final project. Write some code in the below code cell to print out the text of up to five test set examples that your model gets **wrong**. If your model gets more than five test examples wrong, randomly choose five of them to analyze. If your model gets fewer than five examples wrong, please design five test examples that fool your model (i.e., *adversarial examples*). Then, in the following text cell, perform a qualitative analysis of these examples. See if you can figure out any reasons for errors that you observe, or if you have any informed guesses (e.g., common linguistic properties of these particular examples). Does this analysis suggest any possible future steps to improve your classifier?

In [ ]:
## YOUR ERROR ANALYSIS CODE HERE
## print out up to 5 test set examples (or adversarial examples) that your model gets wrong

### *DESCRIBE YOUR QUALITATIVE ANALYSIS OF THE ABOVE EXAMPLES HERE*



---

Finished? Remember to upload the PDF file of this notebook and url of your colab and your three dataset files (annotator1.csv, annotator2.csv, and final_data.csv) to Elearn (elearn.ut.ac.ir) 
